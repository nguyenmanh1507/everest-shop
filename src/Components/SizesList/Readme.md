SizesList example

### Select single

```js
<SizesList name="select-single" handleSizeSelect={() => {}} />
```

### Select multi

```js
<SizesList name="select-multi" selectMulti handleSizeSelect={() => {}} />
```

### Generate sizes of choice

```js
<SizesList
  sizes={['S', 'M', 'L']}
  name="sizes-choice"
  handleSizeSelect={() => {}}
/>
```


