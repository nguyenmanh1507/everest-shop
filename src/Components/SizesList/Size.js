// @flow

import React, { PureComponent } from 'react'

import type { SizeType } from './SizesListType'

class Size extends PureComponent<SizeType> {
  render() {
    const { size, selectMulti, name, handleSizeSelect, ...rest } = this.props

    return (
      <label className="brick-2 mr-xs">
        <input
          data-test="size-input"
          type={selectMulti ? 'checkbox' : 'radio'}
          name={name}
          onChange={handleSizeSelect}
          value={`${name}:${size}`}
          {...rest}
        />
        <span className="brick-2__body">{size}</span>
      </label>
    )
  }
}

export default Size
