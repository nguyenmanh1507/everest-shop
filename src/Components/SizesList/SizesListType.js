// @flow

type SizeSuit = 'XS' | 'S' | 'M' | 'L' | 'XL' | 'XXL'

export type SizesListType = {
  /** Function handle value on change */
  handleSizeSelect: () => void,
  /** Group name for sizes */
  name: string,
  /** Choose multi size or not */
  selectMulti?: boolean,
  /** Array size of choice (size name follow Material sizes) */
  sizes?: Array<SizeSuit>
}

export type SizeType = {
  /** Size name */
  size: string,
  /** Function handle value on change */
  handleSizeSelect: () => void,
  /** Group name for sizes */
  name: string,
  /** Choose multi size or not */
  selectMulti?: boolean
}
