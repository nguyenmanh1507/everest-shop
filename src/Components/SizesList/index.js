// @flow

export { default } from './SizesList'
