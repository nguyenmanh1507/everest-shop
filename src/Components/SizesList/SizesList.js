// @flow

import React, { PureComponent } from 'react'
import { uniqueId, uniq } from 'lodash'

import type { SizesListType } from './SizesListType'
import Size from './Size'

const SizeCodeNames = ['XS', 'S', 'M', 'L', 'XL', 'XXL']

class SizesList extends PureComponent<SizesListType> {
  static defaultProps = {
    selectMulti: false
  }

  render() {
    const { sizes, selectMulti, handleSizeSelect, name, ...rest } = this.props

    if (sizes && sizes.length > 0) {
      if (uniq(sizes).length !== sizes.length) {
        console.warn('Duplicate size name')
      }

      return (
        <div>
          {uniq(sizes).map(size => (
            <Size
              key={uniqueId(`size-${size}-`)}
              size={size}
              selectMulti={selectMulti}
              name={name}
              handleSizeSelect={handleSizeSelect}
              {...rest}
            />
          ))}
        </div>
      )
    }

    return (
      <div>
        {SizeCodeNames.map(size => (
          <Size
            key={uniqueId(`size-${size}-`)}
            size={size}
            selectMulti={selectMulti}
            name={name}
            handleSizeSelect={handleSizeSelect}
            {...rest}
          />
        ))}
      </div>
    )
  }
}

export default SizesList
