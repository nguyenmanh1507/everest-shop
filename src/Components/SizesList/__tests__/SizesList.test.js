// @flow

import React from 'react'
import renderer from 'react-test-renderer'
import { mount, render, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import forEach from 'lodash/forEach'

import SizesList from '../SizesList'

configure({ adapter: new Adapter() })

const mockFn = jest.fn()

test('SizesList render properly', () => {
  const component = renderer.create(
    <SizesList name="select-single-test" handleSizeSelect={() => {}} />
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('SizesList render list of checkbox', () => {
  const component = renderer.create(
    <SizesList
      name="select-single-test"
      selectMulti
      handleSizeSelect={() => {}}
    />
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('SizesList render list of choice sizes', () => {
  const sizes = ['S', 'M', 'L']

  const component = renderer.create(
    <SizesList
      name="select-single-test"
      sizes={sizes}
      handleSizeSelect={() => {}}
    />
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

describe('<ColorList>', () => {
  const colorInputName = 'color'
  const wrapper = render(renderComp(colorInputName))

  test('All input should has same name attribute', () => {
    forEach(wrapper.find('[data-test="size-input"]'), node => {
      expect(node.attribs.name).toBe(colorInputName)
    })
  })

  test('Should render input radio by default', () => {
    forEach(wrapper.find('[data-test="size-input"]'), node => {
      expect(node.attribs.type).toBe('radio')
    })
  })

  test('Should render input radio by default', () => {
    const w = render(renderComp(colorInputName, true))
    forEach(w.find('[data-test="size-input"]'), node => {
      expect(node.attribs.type).toBe('checkbox')
    })
  })

  test('Should has number of input equal sizes props length', () => {
    const sizes = ['S', 'M', 'L']
    const sizesLength = sizes.length
    const w = render(renderComp(colorInputName, undefined, sizes))

    expect(w.find('[data-test="size-input"]')).toHaveLength(sizesLength)
  })

  test('Should remove duplicate name in sizes props', () => {
    const sizes = ['S', 'M', 'L']
    const sizesLength = sizes.length
    const sizesDuplicate = ['S', 'S', 'M', 'L', 'L']
    const w = render(renderComp(colorInputName, undefined, sizesDuplicate))

    expect(w.find('[data-test="size-input"]')).toHaveLength(sizesLength)
  })

  test('Should run handle function once when color select', () => {
    const w = mount(renderComp())
    w
      .find('[data-test="size-input"]')
      .first()
      .simulate('change')
    expect(mockFn.mock.calls.length).toBe(1)
  })
})

function renderComp(
  name = 'select-color-test',
  selectMulti = false,
  sizes = [],
  cb = mockFn
) {
  return (
    <SizesList
      name={name}
      selectMulti={selectMulti}
      sizes={sizes}
      handleSizeSelect={cb}
    />
  )
}
