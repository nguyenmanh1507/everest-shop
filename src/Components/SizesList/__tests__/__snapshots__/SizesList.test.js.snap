// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SizesList render list of checkbox 1`] = `
<div>
  <label
    className="brick-2 mr-xs"
  >
    <input
      data-test="size-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:XS"
    />
    <span
      className="brick-2__body"
    >
      XS
    </span>
  </label>
  <label
    className="brick-2 mr-xs"
  >
    <input
      data-test="size-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:S"
    />
    <span
      className="brick-2__body"
    >
      S
    </span>
  </label>
  <label
    className="brick-2 mr-xs"
  >
    <input
      data-test="size-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:M"
    />
    <span
      className="brick-2__body"
    >
      M
    </span>
  </label>
  <label
    className="brick-2 mr-xs"
  >
    <input
      data-test="size-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:L"
    />
    <span
      className="brick-2__body"
    >
      L
    </span>
  </label>
  <label
    className="brick-2 mr-xs"
  >
    <input
      data-test="size-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:XL"
    />
    <span
      className="brick-2__body"
    >
      XL
    </span>
  </label>
  <label
    className="brick-2 mr-xs"
  >
    <input
      data-test="size-input"
      name="select-single-test"
      onChange={[Function]}
      type="checkbox"
      value="select-single-test:XXL"
    />
    <span
      className="brick-2__body"
    >
      XXL
    </span>
  </label>
</div>
`;

exports[`SizesList render list of choice sizes 1`] = `
<div>
  <label
    className="brick-2 mr-xs"
  >
    <input
      data-test="size-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:S"
    />
    <span
      className="brick-2__body"
    >
      S
    </span>
  </label>
  <label
    className="brick-2 mr-xs"
  >
    <input
      data-test="size-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:M"
    />
    <span
      className="brick-2__body"
    >
      M
    </span>
  </label>
  <label
    className="brick-2 mr-xs"
  >
    <input
      data-test="size-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:L"
    />
    <span
      className="brick-2__body"
    >
      L
    </span>
  </label>
</div>
`;

exports[`SizesList render properly 1`] = `
<div>
  <label
    className="brick-2 mr-xs"
  >
    <input
      data-test="size-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:XS"
    />
    <span
      className="brick-2__body"
    >
      XS
    </span>
  </label>
  <label
    className="brick-2 mr-xs"
  >
    <input
      data-test="size-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:S"
    />
    <span
      className="brick-2__body"
    >
      S
    </span>
  </label>
  <label
    className="brick-2 mr-xs"
  >
    <input
      data-test="size-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:M"
    />
    <span
      className="brick-2__body"
    >
      M
    </span>
  </label>
  <label
    className="brick-2 mr-xs"
  >
    <input
      data-test="size-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:L"
    />
    <span
      className="brick-2__body"
    >
      L
    </span>
  </label>
  <label
    className="brick-2 mr-xs"
  >
    <input
      data-test="size-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:XL"
    />
    <span
      className="brick-2__body"
    >
      XL
    </span>
  </label>
  <label
    className="brick-2 mr-xs"
  >
    <input
      data-test="size-input"
      name="select-single-test"
      onChange={[Function]}
      type="radio"
      value="select-single-test:XXL"
    />
    <span
      className="brick-2__body"
    >
      XXL
    </span>
  </label>
</div>
`;
